Sources/RestKit/RestRequest.swift

```diff
@@ -131,27 +131,28 @@ extension RestRequest {
             // create a task to execute the request
             let task = self.session.dataTask(with: urlRequest) { (data, response, error) in
 
-                // ensure there is no underlying error
-                guard error == nil else {
-                    completionHandler(data, response as? HTTPURLResponse, error)
-                    return
-                }
-
                 // ensure there is a valid http response
                 guard let response = response as? HTTPURLResponse else {
                     let error = RestError.noResponse
                     completionHandler(data, nil, error)
                     return
                 }
 
+                // ensure there is no underlying error
+                guard error == nil else {
+                    let restError = RestError.http(statusCode: response.statusCode, message: "\(String(describing: error))")
+                    completionHandler(data, response, restError)
+                    return
+                }
+
```

Comment:
It works better in this order, since the `error` in the completion handler is no longer a generic `Error` type, but is now a `RestError`. This means that we cannot simply pass back the error we get from the urlRequest, as we were previously. In order to build a `RestError.http`, we need to have a valid response so that we can get the `statusCode`. It is not possible to do this before doing `response = response as? HTTPURLResponse`.

Reply:
Now I understand.  Thx.